client/asset/btc/spv.go

```diff
@@ -465,28 +468,59 @@ func (w *spvWallet) syncHeight() int32 {
 }
 
 // syncStatus is information about the wallet's sync status.
+//
+// The neutrino wallet has a two stage sync:
+//  1. chain service fetching block headers and filter headers
+//  2. wallet address manager retrieving and scanning filters
+//
+// We only report a single sync height, so we are going to simulate some
+// progress in the chain service sync stage that comes before the wallet has
+// performed any address recovery/rescan.
 func (w *spvWallet) syncStatus() (*syncStatus, error) {
-	blk, err := w.cl.BestBlock()
+	// Chain service headers (block and filter) height.
+	chainBlk, err := w.cl.BestBlock()
 	if err != nil {
 		return nil, err
 	}
-
 	target := w.syncHeight()
-	currentHeight := blk.Height
-	synced := w.wallet.ChainSynced()
-	// Sometimes the wallet doesn't report the chain as synced right away.
-	// Seems to be a bug.
-	if !synced && target > 0 && target == currentHeight {
-		synced = true
-	}
-
-	if atomic.SwapInt32(&w.syncTarget, target) == 0 && target > 0 {
-		w.tipChan <- &block{
-			hash:   blk.Hash,
-			height: int64(blk.Height),
+	currentHeight := chainBlk.Height
+	if target == 0 { // definitely not synced no matter what ChainSynced says
+		return &syncStatus{
+			Height:  currentHeight,
+			Syncing: true,
+		}, nil
+	}
+
+	var synced bool
+	var blk *block
+	// Wallet address manager sync height.
+	if currentHeight >= target {
```

Comment:
Note that this condition intentionally prevents `SyncStatus` from reporting `synced = true` before the wallet address manager says so.  So even if progress *may* gets close to 1 in the few seconds while the chain service syncs before consulting the address manager, it will never tell the consumer that it's done.

The ultimate solution might be to get the wallet birthday _block_ from the Wallet instance somehow, and when `w.cl.BestBlock()` is at the birthday block, switch to the address manager's status at that point.  But all I want from this fix is to only report 1.0 and synced=true when the address manager says it has reached the target.

Reply:
The db approach to pulling the birthday _block_ failed because it's not set until the wallet address manager hits the birthday, which makes perfect sense.

All we have in terms of birthday is a `time.Time`.  We should be able to compare the `BlockStamp.Timestamp` with the birthday time to know when to switch over, but the chain service's `BestBlock` does not set `Timestamp`.  This may be resolved in https://github.com/lightninglabs/neutrino/pull/235.  We should circle back on `SyncStatus` when/if that happens.